### PR TITLE
[ASTextNode] Fix deadlock calling trailingRect

### PR DESCRIPTION
**Pull request by maicki**
Fixes issue: #1334

### CODE REVIEW

**Comment by garrettmoon** [comment]
Nice! Fixes the issue and likely even more performant :)

**Comment by maicki** [comment]
@rcancro @garrettmoon Can you guys take another look over this please

File: AsyncDisplayKit/TextKit/ASTextKitRenderer+Positioning.mm [hunk 1]
```diff
@@ -333,22 +333,30 @@ - (void)enumerateTextIndexesAtPosition:(CGPoint)externalPosition usingBlock:(as_
...
     }
-
-    // Take everything after our final character as trailing space.
-    NSArray *finalRects = [self rectsForTextRange:NSMakeRange([textStorage length] - 1, 1) measureOption:ASTextKitRendererMeasureOptionLineHeight];
-    CGRect finalGlyphRect = [[finalRects lastObject] CGRectValue];
-    CGPoint origin = CGPointMake(CGRectGetMaxX(finalGlyphRect), CGRectGetMinY(finalGlyphRect));
-    CGSize size = CGSizeMake(calculatedSize.width - origin.x, calculatedSize.height - origin.y);
-    trailingRect = (CGRect){origin, size};
+      
+    textRange = NSMakeRange([textStorage length] - 1, 1);
```
Comment (garrettmoon) [inline AsyncDisplayKit/TextKit/ASTextKitRenderer+Positioning.mm]: Now that I think about it more, I don't think it's safe to unlock between getting the textRange and calling rectsForTextRange:measureOption:

**Comment by maicki** [comment]
@garrettmoon @rcancro Good for review again. I created a helper function now that does not lock and this can be used if already in a locked state.

**Comment by garrettmoon** [comment]
I think this is the correct solution all though it sucks to have an unlocked and locked version of the method. rectsForTextRange should still behave exactly the same way as before right @maicki?

Looks good to me!

**Comment by rcancro** [comment]
The github diff makes it look like a lot change, but from what I can tell you just moved things around, right? Can you name the unlocked method something really scary sounding so people don't misuse it?

**Comment by maicki** [comment]
@garrettmoon rectsForTextRange works the same way as before now, it just calls the unlocked version of the method.

@rcancro Currently the unlocked version is not exposed and can just be used within the method were it is defined. I hope the comment at the top makes it clear or is enough that developers don't use it without locking.

**Comment by appleguy** [comment]
@rcancro I do agree.  @maicki since this code has changed a lot since I last was familiar with it, it's pretty difficult to understand the change here because of intermixing of pretty complex code segments.  I'm not sure there would have been a way with less change, but particularly for super-core code like this, it might be the one exception to the otherwise-actively-encouraged idea of making cleanup changes as part of all diffs; for locking correctness we should probably keep it to a minimized change, then a different one for improved formatting.

Because @garrettmoon also took a look and I do really trust @maicki 's quality of work + testing, I'm definitely comfortable landing this despite getting a bit lost in the review.

This is a very important fix and definitely a tough, little-known area of the code base.  Thank you @maicki for jumping in to fix it!!

**Pull request merged**